pkg/kv/bulk/sst_batcher.go

```diff
@@ -451,20 +474,66 @@ func (b *SSTBatcher) doFlush(ctx context.Context, reason int) error {
 		b.ms.LastUpdateNanos = timeutil.Now().UnixNano()
 	}
 
-	if err := b.addSSTable(ctx, start, end, b.sstFile.Data()); err != nil {
-		return err
+	stats := b.ms
+	summary := b.rowCounter.BulkOpSummary
+	data := b.sstFile.Data()
+	batchTS := b.batchTS
+	var reserved int64
+
+	fn := func(ctx context.Context) error {
+		if err := b.addSSTable(ctx, batchTS, start, end, data, stats); err != nil {
+			b.mem.Shrink(ctx, reserved)
+			return err
+		}
+		b.mu.Lock()
+		summary.DataSize += int64(size)
+		b.mu.totalRows.Add(summary)
+		atomic.AddInt64(&b.stats.batchWaitAtomic, int64(timeutil.Since(beforeFlush)))
+		atomic.AddInt64(&b.stats.dataSizeAtomic, int64(size))
+		b.mu.Unlock()
+		if reserved != 0 {
+			b.mem.Shrink(ctx, reserved)
+		}
+		return nil
 	}
 
-	b.rowCounter.DataSize += b.sstWriter.DataSize
-	b.totalRows.Add(b.rowCounter.BulkOpSummary)
-	b.stats.batchWait += timeutil.Since(beforeFlush)
-	b.stats.dataSize += sz(b.sstWriter.DataSize)
-	return nil
+	if reason == rangeFlush {
+		// If we're flushing due to a range boundary, we we might be flushing this
+		// one buffer into many different ranges, and doing so one-by-one, waiting
+		// for each round-trip serially, could really add up; a buffer of random
+		// data that covers all of a 2000 range table would be flushing 2000 SSTs,
+		// each of which might be quite small, like 256kib, but still see, say, 50ms
+		// or more round-trip time. Doing those serially would then take minutes. If
+		// we can, instead send this SST and move on to the next while it is sent,
+		// we could reduce that considerably. One concern with doing so however is
+		// that you could potentially end up with an entire buffer's worth of SSTs
+		// all inflight at once, effectively doubling the memory footprint, so we
+		// need to reserve memory from a monitor for the sst before we move on to
+		// the next one; if memory is not available we'll just block on the send
+		// and then move on to the next send after this SST is no longer being held
+		// in memory.
+		if b.asyncAddSSTs == (ctxgroup.Group{}) {
+			b.asyncAddSSTs = ctxgroup.WithContext(ctx)
+		}
+		if err := b.mem.Grow(ctx, int64(cap(data))); err != nil {
+			log.VEventf(ctx, 3, "%s unable to reserve enough memory to flush async: %v", b.name, err)
+		} else {
+			reserved = int64(cap(data))
+			b.asyncAddSSTs.GoCtx(fn)
+			return nil
+		}
+	}
+
+	return fn(ctx)
 }
 
 // Close closes the underlying SST builder.
-func (b *SSTBatcher) Close() {
+func (b *SSTBatcher) Close(ctx context.Context) {
 	b.sstWriter.Close()
+	if err := b.asyncAddSSTs.Wait(); err != nil {
```

Comment:
Swallowing the error is okay because we expect every user of the sstbatcher to flush before closing? Should we return an error or maybe fatal if we see an error here instead, so that incase there is an addsstable request that we have not waited for before, and it fails we don’t just carry on silently.

Reply:
Yeah, if you don’t Flush before Close, it is already the case that anything in sstwriter that hasn’t been sent is lost, so already you have to flush and check its error if you want to know if you wrote everything successfully or not, so this one seems moot; we only do it so we don’t leak gotoutines

Reply:
Makes sense, thanks